File: apps/frontend/evals/chord-detection/metrics.ts
import { SUPPORTED_CHORD_ID_LIST } from "./label-map";
import type {
  ChordVerifierTrial,
  EvaluatedSampleResult,
  MetricsReport,
  PerChordMetrics,
  SampleResult,
} from "./types";
import { computeDurationWeightedMetrics } from "./wcsr";

export function computeMetrics(results: readonly SampleResult[]): MetricsReport {
  const evaluated = results.filter(
    (result): result is EvaluatedSampleResult => result.status === "evaluated",
  );
  const failed = results.length - evaluated.length;
  const perChord = new Map<string, MutableChordMetrics>();
  const confusionMatrix: Record<string, Record<string, number>> = {};
  let topOneCorrect = 0;
  let positiveAccepted = 0;
  let positiveRejected = 0;
  let positiveUncertain = 0;
  let negativeTrials = 0;
  let falseAccepts = 0;
  let wrongAcceptedSamples = 0;

  for (const chordId of SUPPORTED_CHORD_ID_LIST) {
    perChord.set(chordId, { chordId, support: 0, predicted: 0, correct: 0 });
  }

  for (const result of evaluated) {
    const expected = result.expectedChordId;
    const predicted = result.predictedChordId ?? "unknown";
    confusionMatrix[expected] ??= {};
    confusionMatrix[expected][predicted] = (confusionMatrix[expected][predicted] ?? 0) + 1;

    const expectedMetrics = ensureChord(perChord, expected);
    expectedMetrics.support++;
    if (result.correct) topOneCorrect++;

    if (result.verifierStatus === "accepted") {
      positiveAccepted++;
      expectedMetrics.correct++;
      expectedMetrics.predicted++;
    } else if (result.verifierStatus === "rejected") {
      positiveRejected++;
    } else {
      positiveUncertain++;
    }

    const acceptedNegatives = result.negativeTrials.filter(isAccepted);
    negativeTrials += result.negativeTrials.length;
    falseAccepts += acceptedNegatives.length;
    if (acceptedNegatives.length > 0) wrongAcceptedSamples++;
    for (const trial of acceptedNegatives) {
      ensureChord(perChord, trial.expectedChordId).predicted++;
    }
  }

  const evaluatedCount = evaluated.length;
  const weighted = computeDurationWeightedMetrics(evaluated);
  return {
    summary: {
      evaluated: evaluatedCount,
      failed,
      totalDurationSec: weighted.totalDurationSec,
      negativeTrials,
      falseAccepts,
      wrongAcceptedSamples,
      accuracy: safeDivide(topOneCorrect, evaluatedCount),
      verifierRecall: safeDivide(positiveAccepted, evaluatedCount),
      verifierWeightedRecall: weighted.verifierWeightedRecall,
      falseRejectRate: safeDivide(evaluatedCount - positiveAccepted, evaluatedCount),
      falseAcceptRate: safeDivide(falseAccepts, negativeTrials),
      wrongAcceptedRate: safeDivide(wrongAcceptedSamples, evaluatedCount),
      unknownRate: safeDivide(positiveUncertain, evaluatedCount),
      rejectedRate: safeDivide(positiveRejected, evaluatedCount),
      wcsr: weighted.wcsr,
    },
    perChord: [...perChord.values()]
      .filter((item) => item.support > 0 || item.predicted > 0)
      .map(toPerChordMetrics)
      .sort((a, b) => a.chordId.localeCompare(b.chordId)),
    confusionMatrix,
  };
}

function isAccepted(trial: ChordVerifierTrial): boolean {
  return trial.status === "accepted";
}

interface MutableChordMetrics {
  chordId: string;
  support: number;
  predicted: number;
  correct: number;
}

function ensureChord(
  perChord: Map<string, MutableChordMetrics>,
  chordId: string,
): MutableChordMetrics {
  const existing = perChord.get(chordId);
  if (existing) return existing;
  const next = { chordId, support: 0, predicted: 0, correct: 0 };
  perChord.set(chordId, next);
  return next;
}

function toPerChordMetrics(item: MutableChordMetrics): PerChordMetrics {
  const precision = safeDivide(item.correct, item.predicted);
  const recall = safeDivide(item.correct, item.support);
  return {
    ...item,
    precision,
    recall,
    f1: precision + recall === 0 ? 0 : (2 * precision * recall) / (precision + recall),
  };
}

function safeDivide(numerator: number, denominator: number): number {
  return denominator > 0 ? numerator / denominator : 0;
}

File: apps/backend/app/evals/chord_detection/metrics.py
from __future__ import annotations

from .catalog import SUPPORTED_CHORD_ID_LIST
from .wcsr import compute_duration_weighted_metrics


def compute_metrics(results: list[dict]) -> dict:
    evaluated = [result for result in results if result["status"] == "evaluated"]
    failed = len(results) - len(evaluated)
    per_chord = {
        chord_id: {"chordId": chord_id, "support": 0, "predicted": 0, "correct": 0}
        for chord_id in SUPPORTED_CHORD_ID_LIST
    }
    confusion_matrix: dict[str, dict[str, int]] = {}
    top_one_correct = 0
    positive_accepted = 0
    positive_rejected = 0
    positive_uncertain = 0
    negative_trials = 0
    false_accepts = 0
    wrong_accepted_samples = 0

    for result in evaluated:
        expected = result["expectedChordId"]
        predicted = result["predictedChordId"] or "unknown"
        confusion_matrix.setdefault(expected, {})
        confusion_matrix[expected][predicted] = confusion_matrix[expected].get(predicted, 0) + 1
        expected_metrics = ensure_chord(per_chord, expected)
        expected_metrics["support"] += 1
        if result["correct"]:
            top_one_correct += 1
        if result["verifierStatus"] == "accepted":
            positive_accepted += 1
            expected_metrics["correct"] += 1
            expected_metrics["predicted"] += 1
        elif result["verifierStatus"] == "rejected":
            positive_rejected += 1
        else:
            positive_uncertain += 1
        accepted_negatives = [
            trial for trial in result["negativeTrials"] if trial["status"] == "accepted"
        ]
        negative_trials += len(result["negativeTrials"])
        false_accepts += len(accepted_negatives)
        if accepted_negatives:
            wrong_accepted_samples += 1
        for trial in accepted_negatives:
            ensure_chord(per_chord, trial["expectedChordId"])["predicted"] += 1

    evaluated_count = len(evaluated)
    weighted = compute_duration_weighted_metrics(evaluated)
    return {
        "summary": {
            "evaluated": evaluated_count,
            "failed": failed,
            "totalDurationSec": weighted["totalDurationSec"],
            "negativeTrials": negative_trials,
            "falseAccepts": false_accepts,
            "wrongAcceptedSamples": wrong_accepted_samples,
            "accuracy": safe_divide(top_one_correct, evaluated_count),
            "verifierRecall": safe_divide(positive_accepted, evaluated_count),
            "verifierWeightedRecall": weighted["verifierWeightedRecall"],
            "falseRejectRate": safe_divide(evaluated_count - positive_accepted, evaluated_count),
            "falseAcceptRate": safe_divide(false_accepts, negative_trials),
            "wrongAcceptedRate": safe_divide(wrong_accepted_samples, evaluated_count),
            "unknownRate": safe_divide(positive_uncertain, evaluated_count),
            "rejectedRate": safe_divide(positive_rejected, evaluated_count),
            "wcsr": weighted["wcsr"],
        },
        "perChord": [to_per_chord_metrics(item) for item in per_chord.values() if item["support"] > 0 or item["predicted"] > 0],
        "confusionMatrix": confusion_matrix,
    }


def ensure_chord(per_chord: dict[str, dict], chord_id: str) -> dict:
    if chord_id not in per_chord:
        per_chord[chord_id] = {"chordId": chord_id, "support": 0, "predicted": 0, "correct": 0}
    return per_chord[chord_id]


def to_per_chord_metrics(item: dict) -> dict:
    precision = safe_divide(item["correct"], item["predicted"])
    recall = safe_divide(item["correct"], item["support"])
    return {
        **item,
        "precision": precision,
        "recall": recall,
        "f1": 0 if precision + recall == 0 else (2 * precision * recall) / (precision + recall),
    }


def safe_divide(numerator: int | float, denominator: int | float) -> float:
    return float(numerator / denominator) if denominator > 0 else 0.0
